### ISSUE TITLE
Core streamlining

### ISSUE THREAD

**Issue by Zero1627** [Aatu/FieryVoid#351]
Damage dealing procedures were streamlined. As this is already done, I think ittle more work would streamline further relatively common areas.
As notable candidates for streamlining I see:
- Interception overhaul, so it's used more effectively
- Adaptive Armor (now usable just by White Star; also, every special weapon needs to account for it separately...)
- Piercing mode (now usable only as only or second mode; also, I'm not sure if it's handled properly)
- Linked weapons (woud be easy to do with attribute like noOverkill, would avoid serious redefinitions)
- Pulse weapons (also a comon mechanics that uses heavy redefinitions)

**Comment by Zero1627** [comment]
Core firing methods of FV were rewritten pretty completely. Just so You know what’s done, main points of change are:
 
Complete (but not deployed yet):
Core functions now natively undesrtand firing modes (Piercing, Raking, Pulse), eliminating the need of most redefinitions of core functions in particular weapons (and simplifying the rest, allowing to still use the core). Key attributes are new $damageType and $weaponClass.
Core functions now natively understand most common modifiers of firing modes (linked fire, no overkill) as attributes, eliminating the need of still more redefinitions
Standard and special (for now – Adaptive) armor were separated and handled in core functions, simplifying redefinitions
New mechanism for changing firing modes was added (previous ones still do work). It was intended for simplifying mode changes (Raking/Piercing), but hopefully it’s extensive enough to cover almost any change – allowing for easy definitione of complicated multi-weapons (hopefully I’ll be able to create new Laser-Pulse Arrays and Grav Lance, that will actually work as desired)
 
Immediate goals:
Overhaul existing weapons, filling new variables anddisabling or simplifying core overrides (~40% done)
Finish firing mode change mechanism, define appropriate multi-weapons
Deploy and debug all above (before trying for what’s below)…
 
Further goals:
Change allocation of weapons – do it BEFORE any firing is done. I do have an idea how to do it, and it’s worthwhile on its own – but primarily it’s a prerequisite for point 2.
Change interception allocation to something sensible (auto-allocating vs ballistics, too!). My primary goal for a long time now J - essentially everything above is prerequisite for this…

**Comment by tksolway** [comment]
Wow, that’s a massive job.  Great work.

**Comment by Zero1627** [comment]
so it is - but the single most difficult part of first stage is already done. Now lots of minor changes to make entire system consistent... and on to debugging!

**Comment by Zero1627** [comment]
Stage 1 is essentially done - just minor touches remain, and I want to finish a few side things of my own before deploying.

I will deploy only when I have at least a few hours relatively free for initial debugging - I have no delusions about the number of bugs such a deep change must have left in its wake... I want to get out at least the biggest ones immediately after initial deployment.

**Comment by Zero1627** [comment]
deployed but I was too ambitious, left not fully debugged :( , will finish tomorrow.

**Comment by Zero1627** [comment]
OK! I do believe it works!
I do not believe bug-free, but works :) . I declare stage 1 to be complete.

I'll wait some time before starting stage 2 however (so all the bugs may rear their heads and be slain before that).

Goals remaining:
1. Change allocation of weapons – do it BEFORE any firing is done. 
2. Change interception allocation (auto-allocating vs ballistics, too!) - taking into accunt actual hit chance, multi-gun weapons etc.

**Comment by Zero1627** [comment]
Interception redone and working! at last!
